### PR TITLE
nnbuilder-constructor

### PR DESCRIPTION
**Pull request by J3ne4ek**

### CODE REVIEW

File: constructor/constructor.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                ['relu', nn.ReLU()],
+                ['sigmoid', nn.Sigmoid()],
+                ['softmax', nn.Softmax()],
+                ['logsoftmax', nn.LogSoftmax()],
+                ['elu', nn.ELU()],
+    ])
+
+loss_functions = {'MSELoss': torch.nn.MSELoss(), 'CrossEntropy': torch.nn.CrossEntropyLoss()}
+
+class CustomDataset(torch.utils.data.Dataset):
```
Comment (iskhakovt) [inline constructor/constructor.py]: -> Dataset?

File: constructor/constructor.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        loss = model.criterion(prediction, target)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+
+def train(model, dataset, optimizer, history, epochs=100):
+    X_train, X_test, y_train, y_test = dataset.split()
+    for ep in range(epochs):
+        print(f"Epoch {ep + 1} \n-------------------------------")
```
Comment (iskhakovt) [inline constructor/constructor.py]: Use logging: https://docs.python.org/3/library/logging.html

File: constructor/constructor.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    def __len__(self):
+        return len(self.vectors)
+
+    def __getitem__(self, index):
+        return torch.FloatTensor(self.vectors[index]), torch.FloatTensor(self.targets[index])
+
+
+class NeuralNetwork(nn.Module):
+    def __init__(self, info, type):
```
Comment (iskhakovt) [inline constructor/constructor.py]: Move the logic out of constructor, as it will be more complex; just set the fields here

File: constructor/constructor.py (hunk 2 above)
Comment (iskhakovt) [reply constructor/constructor.py]: Will you support progress reporting into db? (makes sense to split in a separate PR)

File: constructor/constructor.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+def test_loop(data, target, model, history):
+    size = len(data)
+    test_loss = 0
+    with torch.no_grad():
+        prediction = model(data).detach()
+        test_loss += model.criterion(prediction, target).item()
+
+    test_loss /= size
+    history.append(test_loss)
```
Comment (iskhakovt) [inline constructor/constructor.py]: What do you need history for?

File: constructor/constructor.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    print(f"Avg loss: {test_loss:>8f}")
+
+
+def save(network, optimizer):
+    dict = network.state_dict()
+    tensor_to_lists(dict)
+    state = {
+        'state_dict': dict
```
Comment (iskhakovt) [inline constructor/constructor.py]: This should be a static const

File: constructor/constructor.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        else:
+            self.criterion = loss_functions['MSELoss']
+
+    def forward(self, x):
+        return self.layers(x)
+
+
+def tensor_to_lists(dictionary):
+    for key, value in dictionary.items():
+        dictionary[key] = value.tolist()
```
Comment (iskhakovt) [inline constructor/constructor.py]: Why not constructing a new dict?

File: constructor/constructor.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+
+def list_to_tensor(dictionary):
+    for key, value in dictionary.items():
+        dictionary[key] = torch.Tensor(value)
```
Comment (iskhakovt) [inline constructor/constructor.py]: Why not constructing a new dict?

File: constructor/constructor.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+def train_loop(data, target, model, optimizer):
+        prediction = model(data)
+        loss = model.criterion(prediction, target)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+
+def train(model, dataset, optimizer, history, epochs=100):
```
Comment (iskhakovt) [inline constructor/constructor.py]: 100 should be a static const

File: constructor/constructor.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+def restore_net(states, model):
+    state_dict = json.loads(states)
+    dict = state_dict['state_dict']
+    list_to_tensor(dict)
+    model.load_state_dict(dict)
+    return model
+
+
+def get_prediction(network, x, type='classification'):
```
Comment (iskhakovt) [inline constructor/constructor.py]: There's no need for default type here

File: constructor/constructor.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    return prediction.data.numpy()
+
+
+def accuracy(model, data, target):
+    total = 0
+    correct = 0
+    for X, y in (zip(data, target)):
+        res = model(X)
+        total += 1
+        correct += (abs(res - y) <= 0.5).sum()
```
Comment (iskhakovt) [inline constructor/constructor.py]: what's 0.5 here?

File: constructor/constructor.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        return torch.FloatTensor(self.vectors[index]), torch.FloatTensor(self.targets[index])
+
+
+class NeuralNetwork(nn.Module):
+    def __init__(self, info, type):
+        super(NeuralNetwork, self).__init__()
+        self.array = []
+        layers = info['layers']
+        for i in range(len(layers) - 1):
+            self.array.append(nn.Linear(layers[i]['neurons'], layers[i + 1]['neurons']))
```
Comment (iskhakovt) [inline constructor/constructor.py]: Have you added a free var?

File: constructor/structures.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    ord_enc = OrdinalEncoder()
+    col = df.select_dtypes(include=['object']).columns
+    df[col] = ord_enc.fit_transform(df[col])
+
+
+class Dataset(object):
+    def __init__(self, args):
+        data_bytes, target_name = args[1:]
+        f = io.BytesIO(data_bytes)
+        df = pd.read_csv(f)
```
Comment (iskhakovt) [inline constructor/structures.py]: In general, you shouldn't load the whole ds into memory at once, as it may be too huge. Use batching and stochastic gradient descent.

File: constructor/train.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import json
+import torch
+import structures
+import random
+from psycopg2 import sql
+
+curr_id = int(sys.argv[1])
+
+con = psycopg2.connect(
+  database="nnbuilder",
```
Comment (iskhakovt) [inline constructor/train.py]: This shouldn't be hardcoded, use environment variables.

File: constructor/train.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  user="nnbuilder",
+  password="nnbuilder",
+  host="127.0.0.1",
+  port="5432"
+)
+
+cur = con.cursor()
+
+cur.execute("SELECT * from tasksqueue where task_id=%s", [curr_id])
```
Comment (iskhakovt) [inline constructor/train.py]: from -> FROM, where -> WHERE, ... here and below

File: constructor/train.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+import io
+import constructor as ctor
+import psycopg2
+import json
+import torch
+import structures
+import random
+from psycopg2 import sql
+
+curr_id = int(sys.argv[1])
```
Comment (iskhakovt) [inline constructor/train.py]: This name isn't descriptive

File: constructor/train.py (hunk 14 above)
Comment (iskhakovt) [reply constructor/train.py]: You use f'...{}' above, that's inconsistent

File: constructor/train.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    cur.execute("SELECT * from models where m_id=%s", [task_info.model])
+    model_params = cur.fetchone()
+    ctor.restore_net(model_params[1], model)
+
+if task_info.task_type == structures.TaskType.applyToData:
+    res = ctor.get_prediction(model, dataset.data)
+    df = pd.DataFrame(dict(label=res))
+    data = df.to_csv(index=False)
+
+    id_value = random.randint(0, 10000000)
```
Comment (iskhakovt) [inline constructor/train.py]: What's this constant about?
Also, use numpy random instead

File: constructor/train.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+cur.execute("SELECT * from datasets where ds_id=%s", [task_info.dataset])
+row = cur.fetchone()
+dataset = structures.Dataset(row)
+cur.execute("SELECT * from neuralnetworks where nn_id=%s", [task_info.nnDescription])
+row = cur.fetchone()
+
+info = json.loads(row[1])
+model = ctor.NeuralNetwork(info, task_info.task_type)
+
+if task_info.model is not None:
```
Comment (iskhakovt) [inline constructor/train.py]: What happens if it's none?

**Review by iskhakovt — CHANGES_REQUESTED** [review]

File: constructor/project_sample-3.ipynb [hunk 18]
```diff
@@ -0,0 +1,705 @@
+{
```
Comment (iskhakovt) [inline constructor/project_sample-3.ipynb]: Remove this file

File: constructor/database_tests.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+def generate_task_in_db(nn_id_, dataset_id_):
+    id_value = random.randint(0, 10000000)
+
+    cur.execute(
+        "INSERT INTO tasksqueue VALUES (%s, %s, %s);", (id_value, nn_id_, dataset_id_)
+    )
+
+
+con = psycopg2.connect(
```
Comment (iskhakovt) [inline constructor/database_tests.py]: Duplicated code

File: constructor/train.py (hunk 17 above)
Comment (J3ne4ek) [reply constructor/train.py]: This means that the model has not been trained before and the default states

File: constructor/constructor.py (hunk 4 above)
Comment (J3ne4ek) [reply constructor/constructor.py]: Now I save it  with model in db for report progress

**Review by iskhakovt — APPROVED** [review]
Not tested to build
